On OSX, with Clang, C++ backend fails to compile when using closure type in hashset or as table key.

On OSX, with Clang, C++ backend fails to compile when using closure type in hashset or as table key.

### Example:
```nim
import sets

type callType = proc() {.closure.}
var setA = initHashSet[callType]()

let foo = proc()= discard
setA.incl(foo)
doAssert setA.contains(foo)
```

Error message from c++ backend is as below:
```
error: assigning to 'void *' from incompatible type 'void (*)(void *)'
             result = x.ClP_0;
                      ~~^~~~~
```
Tested in both 1.4.0 and the devel version, on OSX Mojave.


```
$ nim -v
Nim Compiler Version 1.5.1 [MacOSX: amd64]
Compiled at 2020-11-22
Copyright (c) 2006-2020 by Andreas Rumpf

git hash: f1764aaa9ed2a4e3488ccc5632aed77e3cae8cbf
active boot switches: -d:release
```

```
$clang -v  
Apple clang version 11.0.0 (clang-1100.0.33.17)
Target: x86_64-apple-darwin18.7.0
Thread model: posix
```

Can reproduce on Linux with clang 11.1.0:
```
CC: stdlib_sets.nim
/home/dian/.cache/nim/asd_d/stdlib_sets.nim.cpp:380:19: error: assigning to 'void *' from 'void (*)(void *)' converts between void pointer and function pointer
              result = x.ClP_0;
                       ~~^~~~~
1 error generated.
Error: execution of an external compiler program 'clang++ -c -std=gnu++14 -funsigned-char    -I/home/dian/Things/Nim/lib -I/home/dian/Projects/refen -o /home/dian/.cache/nim/asd_d/stdlib_sets.nim.cpp.o /home/dian/.cache/nim/asd_d/stdlib_sets.nim.cpp' failed with exit code: 1
```

--- a/tests/sets/tsets_various.nim
+++ b/tests/sets/tsets_various.nim
@@ -276,10 +276,9 @@ template main() =
   block: # bug #16123
     whenVMorJs: discard
     do:
-      type callType = proc() {.closure.}
-      var setA = initHashSet[callType]()
-
-      let foo = proc()= discard
+      type CallType = proc() {.closure.}
+      var setA = initHashSet[CallType]()
+      let foo = proc() = discard
       setA.incl(foo)
       doAssert setA.contains(foo)
 

--- a/tests/sets/tsets_various.nim
+++ b/tests/sets/tsets_various.nim
@@ -1,15 +1,12 @@
 discard """
   targets: "c cpp js"
-  output: '''
-set is empty
-'''
 """
-
 
 import std/[sets, hashes]
 
 from std/sequtils import toSeq
 from std/algorithm import sorted
+from stdtest/testutils import whenVMorJs
 
 proc sortedPairs[T](t: T): auto = toSeq(t.pairs).sorted
 template sortedItems(t: untyped): untyped = sorted(toSeq(t))
@@ -23,10 +20,12 @@ block tsetpop:
     discard a.pop()
   doAssert len(a) == 0
 
+  var msg = ""
   try:
     echo a.pop()
   except KeyError as e:
-    echo e.msg
+    msg = e.msg
+  doAssert msg == "set is empty"
 
 
 
@@ -258,6 +257,7 @@ block: # test correctness after a number of inserts/deletes
 
 
 template main() =
+  # xxx move all tests inside this
   block:
     let a = {true, false}
     doAssert $a == "{false, true}"
@@ -273,6 +273,15 @@ template main() =
     doAssert $a == "{false}"
     doAssert a.len == 1
 
+  block: # bug #16123
+    whenVMorJs: discard
+    do:
+      type callType = proc() {.closure.}
+      var setA = initHashSet[callType]()
+
+      let foo = proc()= discard
+      setA.incl(foo)
+      doAssert setA.contains(foo)
 
 static: main()
 main()

